### Ch4_Q13
This question should be answered using the Weekly data set, which is part of the ISLP package. <br>
This data is similar in nature to the Smarket data from this chapter’s lab, except that it contains 1,089 weekly <br>
returns for 21 years, from the beginning of 1990 to the end of 2010.

In [69]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from statsmodels.api import Logit
from statsmodels.tools import add_constant

In [2]:
!pip install ISLP

In [70]:
from ISLP import load_data
df = load_data('Weekly')
print(df)

      Year   Lag1   Lag2   Lag3   Lag4   Lag5    Volume  Today Direction
0     1990  0.816  1.572 -3.936 -0.229 -3.484  0.154976 -0.270      Down
1     1990 -0.270  0.816  1.572 -3.936 -0.229  0.148574 -2.576      Down
2     1990 -2.576 -0.270  0.816  1.572 -3.936  0.159837  3.514        Up
3     1990  3.514 -2.576 -0.270  0.816  1.572  0.161630  0.712        Up
4     1990  0.712  3.514 -2.576 -0.270  0.816  0.153728  1.178        Up
...    ...    ...    ...    ...    ...    ...       ...    ...       ...
1084  2010 -0.861  0.043 -2.173  3.599  0.015  3.205160  2.969        Up
1085  2010  2.969 -0.861  0.043 -2.173  3.599  4.242568  1.281        Up
1086  2010  1.281  2.969 -0.861  0.043 -2.173  4.835082  0.283        Up
1087  2010  0.283  1.281  2.969 -0.861  0.043  4.454044  1.034        Up
1088  2010  1.034  0.283  1.281  2.969 -0.861  2.707105  0.069        Up

[1089 rows x 9 columns]


#### Recall (d)

In [71]:
train_data = df[df['Year'] <= 2008]
test_data = df[df['Year'] > 2008]
X_train, y_train = train_data[['Lag2']], train_data['Direction']
X_test, y_test = test_data[['Lag2']], test_data['Direction']

#### (e) Repeat (d) using LDA.

In [72]:
lda_model = LinearDiscriminantAnalysis().fit(X_train, y_train)
y_pred_lda = lda_model.predict(X_test)
conf_matrix_e = confusion_matrix(y_test, y_pred_lda)
accuracy_e = accuracy_score(y_test, y_pred_lda)

# Compute the confusion matrix and accuracy
print('Confusion Matrix:')
df_LDA = pd.DataFrame(conf_matrix_e, 
                    index=['Actual Down', 'Actual Up'], 
                    columns=['Predicted Down', 'Predicted Up'])
print(df_LDA)
print(f"LDA_Accuracy: {accuracy_e:.4f}")

Confusion Matrix:
             Predicted Down  Predicted Up
Actual Down               9            34
Actual Up                 5            56
LDA_Accuracy: 0.6250


#### (f) Repeat (d) using QDA.

In [73]:
qda_model = QuadraticDiscriminantAnalysis().fit(X_train, y_train)
y_pred_qda = qda_model.predict(X_test)
conf_matrix_f = confusion_matrix(y_test, y_pred_qda)
accuracy_f = accuracy_score(y_test, y_pred_qda)

print('Confusion Matrix:')
df_QDA = pd.DataFrame(conf_matrix_f, 
                    index=['Actual Down', 'Actual Up'], 
                    columns=['Predicted Down', 'Predicted Up'])
print(df_QDA)
print(f"QDA_Accuracy: {accuracy_f:.4f}")

Confusion Matrix:
             Predicted Down  Predicted Up
Actual Down               0            43
Actual Up                 0            61
QDA_Accuracy: 0.5865


#### (g) Repeat (d) using KNN with K = 1.

In [74]:
knn_model = KNeighborsClassifier(n_neighbors=1).fit(X_train, y_train)
y_pred_knn = knn_model.predict(X_test)
conf_matrix_g = confusion_matrix(y_test, y_pred_knn)
accuracy_g = accuracy_score(y_test, y_pred_knn)

print('Confusion Matrix:')
df_KNN = pd.DataFrame(conf_matrix_g, 
                    index=['Actual Down', 'Actual Up'], 
                    columns=['Predicted Down', 'Predicted Up'])
print(df_KNN)
print(f"KNN with K=1_Accuracy: {accuracy_g:.4f}")

Confusion Matrix:
             Predicted Down  Predicted Up
Actual Down              22            21
Actual Up                31            30
KNN with K=1_Accuracy: 0.5000


#### (h) Repeat (d) using naive Bayes.

In [75]:
nb_model = GaussianNB().fit(X_train, y_train)
y_pred_nb = nb_model.predict(X_test)
conf_matrix_h = confusion_matrix(y_test, y_pred_nb)
accuracy_h = accuracy_score(y_test, y_pred_nb)

print('Confusion Matrix:')
df_Bayes = pd.DataFrame(conf_matrix_h, 
                    index=['Actual Down', 'Actual Up'], 
                    columns=['Predicted Down', 'Predicted Up'])
print(df_Bayes)
print(f"naive Bayes_Accuracy: {accuracy_h:.4f}")

Confusion Matrix:
             Predicted Down  Predicted Up
Actual Down               0            43
Actual Up                 0            61
naive Bayes_Accuracy: 0.5865


#### (i) Which of these methods appears to provide the best results on this data?

Both Logistic Regression in part(d) and LDA in part(e) provide the best results since their accuracy are the same is 0.625.

(j) Experiment with different combinations of predictors, including possible transformations and interactions, for each of the methods.<br>Report the variables, method, and associated confusion matrix that appears to provide the best results on the held out data.<br>Note that you should also experiment with values for K in the KNN classifier.

Suggests experimenting with combinations of predictors, transformations, and K values for KNN. This part can be customized further depending on the outcomes and available computational resources.

In [76]:
# Define possible feature columns and initialize a dictionary to store results
features = ['Lag1', 'Lag2', 'Lag3', 'Lag4', 'Lag5', 'Volume']
results = {}

# Experiment with different combinations of features and models
for n in range(1, len(features) + 1):
    for combo in combinations(features, n):
        X_train = train_data[list(combo)]
        X_test = test_data[list(combo)]
        
        # Logistic Regression
        logit_model = LogisticRegression().fit(X_train, y_train)
        y_pred_logit = logit_model.predict(X_test)
        accuracy_logit = accuracy_score(y_test, y_pred_logit)
        results[(combo, 'Logistic Regression')] = accuracy_logit
        print(f"Logistic Regression with features {combo}: Accuracy = {accuracy_logit:.4f}")

        # LDA
        lda_model = LinearDiscriminantAnalysis().fit(X_train, y_train)
        y_pred_lda = lda_model.predict(X_test)
        accuracy_lda = accuracy_score(y_test, y_pred_lda)
        results[(combo, 'LDA')] = accuracy_lda
        print(f"LDA with features {combo}: Accuracy = {accuracy_lda:.4f}")

        # QDA
        qda_model = QuadraticDiscriminantAnalysis().fit(X_train, y_train)
        y_pred_qda = qda_model.predict(X_test)
        accuracy_qda = accuracy_score(y_test, y_pred_qda)
        results[(combo, 'QDA')] = accuracy_qda
        print(f"QDA with features {combo}: Accuracy = {accuracy_qda:.4f}")

        # KNN with different values of K
        for k in range(1, 6):  # Try K from 1 to 5
            knn_model = KNeighborsClassifier(n_neighbors=k).fit(X_train, y_train)
            y_pred_knn = knn_model.predict(X_test)
            accuracy_knn = accuracy_score(y_test, y_pred_knn)
            results[(combo, f'KNN (K={k})')] = accuracy_knn
            print(f"KNN (K={k}) with features {combo}: Accuracy = {accuracy_knn:.4f}")

        # Naive Bayes
        nb_model = GaussianNB().fit(X_train, y_train)
        y_pred_nb = nb_model.predict(X_test)
        accuracy_nb = accuracy_score(y_test, y_pred_nb)
        results[(combo, 'Naive Bayes')] = accuracy_nb
        print(f"Naive Bayes with features {combo}: Accuracy = {accuracy_nb:.4f}")

# Find the best result
best_combo, best_model = max(results, key=results.get)
best_accuracy = results[(best_combo, best_model)]
print(f"\nBest Model: {best_model} with features {best_combo}")
print(f"Best Accuracy: {best_accuracy:.4f}")

# Display the confusion matrix for the best model
if 'KNN' in best_model:
    k = int(best_model.split('=')[1].replace(')', ''))
    best_model_instance = KNeighborsClassifier(n_neighbors=k).fit(train_data[list(best_combo)], y_train)
elif best_model == 'Logistic Regression':
    best_model_instance = LogisticRegression().fit(train_data[list(best_combo)], y_train)
elif best_model == 'LDA':
    best_model_instance = LinearDiscriminantAnalysis().fit(train_data[list(best_combo)], y_train)
elif best_model == 'QDA':
    best_model_instance = QuadraticDiscriminantAnalysis().fit(train_data[list(best_combo)], y_train)
elif best_model == 'Naive Bayes':
    best_model_instance = GaussianNB().fit(train_data[list(best_combo)], y_train)


# Compute the confusion matrix and accuracy
print('Confusion Matrix:')
df_best = pd.DataFrame(conf_matrix_best, 
                    index=['Actual Down', 'Actual Up'], 
                    columns=['Predicted Down', 'Predicted Up'])
print(df_best)

Logistic Regression with features ('Lag1',): Accuracy = 0.5673
LDA with features ('Lag1',): Accuracy = 0.5673
QDA with features ('Lag1',): Accuracy = 0.5865
KNN (K=1) with features ('Lag1',): Accuracy = 0.4615
KNN (K=2) with features ('Lag1',): Accuracy = 0.4519
KNN (K=3) with features ('Lag1',): Accuracy = 0.4327
KNN (K=4) with features ('Lag1',): Accuracy = 0.4423
KNN (K=5) with features ('Lag1',): Accuracy = 0.4712
Naive Bayes with features ('Lag1',): Accuracy = 0.5865
Logistic Regression with features ('Lag2',): Accuracy = 0.6250
LDA with features ('Lag2',): Accuracy = 0.6250
QDA with features ('Lag2',): Accuracy = 0.5865
KNN (K=1) with features ('Lag2',): Accuracy = 0.5000
KNN (K=2) with features ('Lag2',): Accuracy = 0.4712
KNN (K=3) with features ('Lag2',): Accuracy = 0.5481
KNN (K=4) with features ('Lag2',): Accuracy = 0.5769
KNN (K=5) with features ('Lag2',): Accuracy = 0.5385
Naive Bayes with features ('Lag2',): Accuracy = 0.5865
Logistic Regression with features ('Lag3',): A